File: project_euler/p004/solution.hs
-- A palindromic number reads the same both ways. The 
-- largest palindrome made from the product of two 
-- 2-digit numbers is 9009 = 91 × 99.

-- Find the largest palindrome made from the product 
-- of two 3-digit numbers.

module Main where

isPalindrome :: Int -> Bool 
isPalindrome ls = (reverse $ show ls) == (show ls)

getMax :: Int -> Int -> Int
getMax bound1 bound2 = foldl1 max [a * b | 
                                   a <- [bound1..bound2], 
                                   b <- [bound1..bound2], 
                                   isPalindrome(a * b)]

main = do print (getMax 100 999)

File: project_euler/p004/solution.rs
/*
A palindromic number reads the same both ways. The 
largest palindrome made from the product of two 
2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product 
of two 3-digit numbers.
*/

fn is_palindrome(num: String) -> bool {

    //Zip the number with the reverse of itself
    let mut z = num.chars().zip(num.chars().rev());

    //For each element in z, check that the the pair equals itself
    z.all(|(first, second)| first == second)
}

fn main() {

    let mut max = 0;

    for i in 100..=1000 {
        for j in 100..=1000 {

            //Compute product
            let product: isize = i*j;

            //As a string
            let as_string: String = product.to_string();

            if is_palindrome(as_string) && product > max {
                max = product
            }
        }
    }

    println!("{max}")
}